### PR TITLE
Shapes interpolation

### PR DESCRIPTION
**Pull request by brisvag**
# Description
Picking up the work from #1256 and #2345. There are a few improvements over the previous implementation:

1. order of interpolation is settable (instead of fixed at cubic)
2. closed polygons are interpolated periodically, which prevents the kink between the first and last point
3. some small stuff, such as number of samples changing based on how many "raw" points are passed

https://user-images.githubusercontent.com/23482191/201670962-21ed6ba7-a83f-474c-8179-19b1ba2499b9.mp4

None of this is currently exposed (we could even merge this as is); I waited to expose it cause I'm not sure what's the best way. Should the whole layer have an interpolation order which is propagated to all its contents? Should one be able to enable/change the interpolation of a single shape? Or maybe we should simply have extra buttons for `interpolated polygon` and `interpolated path` as suggested in #2345 (though this would make it harder to allow changing interpolation order)?

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

File: napari/layers/shapes/_shapes_models/_polgyon_base.py [hunk 1]
```diff
@@ -38,6 +41,8 @@ def __init__(
         filled=True,
         closed=True,
         name='polygon',
+        interpolation_order=1,
```
Comment (brisvag) [inline napari/layers/shapes/_shapes_models/_polgyon_base.py:44]: If you want to test locally, you can change this to 3.

**Comment by psobolewskiPhD** [comment]
Wow that looks awesome.
I'm not sure how to surface this.
I'm not sure it should apply to existing shapes, but I can imagine wanting to mix and match on the same Shapes layer when creating new shapes...
Shapes already has a lot of buttons. Maybe how alternate tools are shown in Adobe apps and ImageJ, where the corner is marked and one can right-click to change the tool? Here's ImageJ example:
image
Then the spline versions would be provided as alternate?

**Comment by brisvag** [comment]
Pinging our resident design people (we really need a pingable group for this...) @isabela-pf, @ppwadhwa.

File: napari/layers/shapes/_shapes_models/_polgyon_base.py (hunk 1 above)
Comment (sofroniewn) [reply napari/layers/shapes/_shapes_models/_polgyon_base.py:44]: nice - so 1 means off by default

**Review by sofroniewn — APPROVED** [review]
Functionality looks great - let's get this in (which is no change as off by default) - and then can expose parameter on layer and in GUI 🚀

**Review by sofroniewn — COMMENTED** [review]
You could add a test actually - that might be nice

File: napari/layers/shapes/_shapes_models/_polgyon_base.py (hunk 1 above)
Comment (brisvag) [reply napari/layers/shapes/_shapes_models/_polgyon_base.py:44]: Yup!

**Comment by brisvag** [comment]
Added some simple tests with hardcoded numbers, just to make sure that *something* happens. Not sure if there's a smart/sensible way of testing that the actual interpolation looks correct...

**Comment by brisvag** [comment]
@sofroniewn bump for a quick check!

**Comment by jni** [comment]
@brisvag I haven't checked the code but the `interpolation_order=1` default makes me wonder whether there is a performance cost here for the "no change" use case? Could you reassure me that there isn't before merging? 😅

**Comment by jni** [comment]
(But if there isn't feel free to merge, and thank you *very much* for reviving this! ❤️)

**Review by sofroniewn — APPROVED** [review]
LGTM!

File: napari/layers/shapes/_shapes_models/_polgyon_base.py [hunk 2]
```diff
@@ -77,10 +85,37 @@ def data(self, data):
...
+        # when a point was just created and the new potential point is set to exactly the same
+        # to prevent issues, we remove the extra points.
+        duplicates = np.isclose(data, np.roll(data, 1, axis=0))
+        # cannot index with bools directly (flattens by design)
+        data_spline = data[~np.all(duplicates, axis=1)]
+
+        if (
+            self.interpolation_order > 1
+            and len(data_spline) > self.interpolation_order
+        ):
```
Comment (brisvag) [inline napari/layers/shapes/_shapes_models/_polgyon_base.py:101]: @jni no cost, new code is all hidden behind this :)

**Pull request merged**